NEW: DolDeprecationHandler for deprecations

mdeweerd opened the pull request:
# NEW: DolDeprecationHandler for deprecations
    
This reusable class avoids needed to implement double assignments while enabling
the detection of the use of deprecated variables and methods.
    
There is no efficiency overhead when the proper variables and methods are used.

Applied on CommonObject class (separate commit).
Fixed PHPUnit testing template as well (separate commit)

mdeweerd committed:
NEW: DolDeprecationHandler for deprecations

# NEW: DolDeprecationHandler for deprecations

This reusable class avoids needed to implement double assignments while enabling
the detection of the use of deprecated variables and methods.

There is no efficiency overhead when the proper variables and methods are
used.

mdeweerd committed:
New: Use DolDeprecationHandler in CommonObject

# New: Use DolDeprecationHandler in CommonObject

Use deprecation handler for more efficient deprecations

eldy commented on htdocs/modulebuilder/template/test/phpunit/MyObjectTest.php:
```diff
@@ -81,7 +82,7 @@ public function __construct($name = '')
 	 * @return void
 	 */
-	public static function setUpBeforeClass()
+	public static function setUpBeforeClass(): void
```
As the transition from PHP7.0 to PHP7.1+ is not yet complete by a lot of users, for the moment, we must avoid the use of the 
": type"
We will continue to wait a little bit that the nb of users with php 7.0 is lower (A lot of users first upgrade their Dolibarr to be able to upgrade their PHP, so having a Dolibarr not hanging on fatal error with PHP 7.0 is still important) .

mdeweerd commented:
@eldy 

I think this is a positive side-effect of the DeprecationHandler: it detects access to undefined properties.

What do you suggest?  I could ignore those for now, but I think it's better for the project to fix them.

```
1) NumberingModulesTest::testFactureMercure
Undefined property 'ventil'
/home/travis/build/Dolibarr/dolibarr/htdocs/core/class/doldeprecationhandler.class.php:66
/home/travis/build/Dolibarr/dolibarr/htdocs/compta/facture/class/facture.class.php:3978
/home/travis/build/Dolibarr/dolibarr/htdocs/compta/facture/class/facture.class.php:843
/home/travis/build/Dolibarr/dolibarr/test/phpunit/NumberingModulesTest.php:164
```

fappels commented:
@mdeweerd Is it compatible with IDE plugin like PHP Intelephense who lookup the @deprecated ?

mdeweerd commented:
> @mdeweerd Is it compatible with IDE plugin like PHP Intelephense who lookup the ATdeprecated ?
I do not use intelephense, but the deprecations are still documented as deprecated properties in PHPDoc in the class where they used to exist (currently only CommonObject) and my tools can work with that information.  So I guess that Intelephense does too.
I use another PHP framewhere where all DB properties are "virtual" and only known to the IDE because of such property definitions.

mdeweerd committed:
Fix: note<->note_private aliasing is handled by DeprecationHandler

# Fix: note<->note_private aliasing is handled by DeprecationHandler

The extra copy is useless, the aliasing for the deprecated property is already handled

frederic34 commented:
@mdeweerd is it possible to keep deprecated and put them private instead public to force use of magic getter ans setter ?

mdeweerd committed:
Fix: Check that certain properties exists before trying isset.

# Fix: Check that certain properties exists before trying isset.

Avoids triggering an undefined property notice from the DolDeprecationHandler

mdeweerd commented:
> is it possible to keep deprecated and put them private instead public to force use of magic getter ans setter ?

Good idea, I implemented a test case and it works there ;-).

mdeweerd committed:
New: Apply DolDeprecatedHandler to Conf

# New: Apply DolDeprecatedHandler to Conf

Improve detection of deprecated values, avoid duplicating assignments,
alias deprecated property names to new properties.

mdeweerd commented:
I made the deprecated properties private.

I managed to run unit tests on my Windows machine (Wamp/cygwin).
However the database setup does not seem perfect (script: https://github.com/mdeweerd/dolibarr/blob/core/deprecationclass/dev/setup/phpunit/setup_conf.sh ).
The datestartvalidity is missing in llx_user and prevents me from logging in (unless it's the password - but I tried demo/demo as well).
![image](https://github.com/Dolibarr/dolibarr/assets/1504752/199c458d-e1d9-441e-9bf1-db7b713570b8)
MariaDB:3306/travis/llx_user/		http://localhost/phpmyadmin/index.php?route=/table/sql&db=travis&table=llx_user
La requête SQL a été exécutée avec succès.
```plaintext
describe `llx_user`;

rowid	int(11)	NO	PRI	NULL	auto_increment	
datec	datetime	YES		NULL		
tms	timestamp	NO		current_timestamp()	on update current_timestamp()	
fk_user_creat	int(11)	YES		NULL		
fk_user_modif	int(11)	YES		NULL		
login	varchar(50)	NO	MUL	NULL		
entity	int(11)	NO		1		
civility	varchar(6)	YES		NULL		
ref_ext	varchar(50)	YES		NULL		
ref_int	varchar(50)	YES		NULL		
employee	smallint(6)	YES		1		
fk_establishment	int(11)	YES		0		
pass	varchar(128)	YES		NULL		
pass_crypted	varchar(128)	YES		NULL		
pass_temp	varchar(128)	YES		NULL		
api_key	varchar(128)	YES	MUL	NULL		
lastname	varchar(50)	YES		NULL		
firstname	varchar(50)	YES		NULL		
job	varchar(128)	YES		NULL		
office_phone	varchar(20)	YES		NULL		
office_fax	varchar(20)	YES		NULL		
user_mobile	varchar(20)	YES		NULL		
email	varchar(255)	YES		NULL		
signature	text	YES		NULL		
admin	smallint(6)	YES		0		
webcal_login	varchar(25)	YES		NULL		
module_comm	smallint(6)	YES		1		
module_compta	smallint(6)	YES		1		
fk_soc	int(11)	YES	MUL	NULL		
fk_socpeople	int(11)	YES	UNI	NULL		
fk_member	int(11)	YES	UNI	NULL		
note_private	text	YES		NULL		
datelastlogin	datetime	YES		NULL		
datepreviouslogin	datetime	YES		NULL		
egroupware_id	int(11)	YES		NULL		
ldap_sid	varchar(255)	YES		NULL		
statut	tinyint(4)	YES		1		
photo	varchar(255)	YES		NULL		
lang	varchar(6)	YES		NULL		
openid	varchar(255)	YES		NULL		
fk_user	int(11)	YES		NULL		
thm	double(24,8)	YES		NULL		
address	varchar(255)	YES		NULL		
zip	varchar(25)	YES		NULL		
town	varchar(50)	YES		NULL		
fk_state	int(11)	YES		0		
fk_country	int(11)	YES		0		
color	varchar(6)	YES		NULL		
accountancy_code	varchar(32)	YES		NULL		
barcode	varchar(255)	YES		NULL		
fk_barcode_type	int(11)	YES		0		
nb_holiday	int(11)	YES		0		
salary	double(24,8)	YES		NULL		
birth	date	YES		NULL		
dateemployment	date	YES		NULL		
dateemploymentend	date	YES		NULL		
default_range	int(11)	YES		NULL		
default_c_exp_tax_cat	int(11)	YES		NULL		
tjm	double(24,8)	YES		NULL		
salaryextra	double(24,8)	YES		NULL		
weeklyhours	double(16,8)	YES		NULL		
gender	varchar(10)	YES		NULL		
note_public	text	YES		NULL		
birth_place	varchar(64)	YES		NULL		
flagdelsessionsbefore	datetime	YES		NULL		
email_oauth2	varchar(255)	YES		NULL		
```

mdeweerd commented:
I managed to setup the demo database.

On `http://127.0.1.0/product/stock/list.php?leftmenu=` (warehouse/stock), we can see that several "dynamic" properties are set, notified by the DolDeprecationHandler:

![image](https://github.com/Dolibarr/dolibarr/assets/1504752/f5cf5609-980b-442c-8c89-cf4fe5583853)

![image](https://github.com/Dolibarr/dolibarr/assets/1504752/3b21d3b8-bbb3-45cb-bbed-05909bd4e6ce)

This revealed that quite a few "isset()" tests are done on properties that do not exist, and there are also dynamic variables being set on CommonObject.  So I made DolDeprecationHandler a bit more forgiving for those cases.

mdeweerd commented:
@eldy On my Windows platform a lot of the file tests fail because of the open_basedir restriction which I set to the development directory - any invalid directory trigers the error notice making the execution of the test stop.

Is it ok if I prefix all the functions that trigger this (@is_dir, and othre functions) - I just want to prefiix them all with '@' - maybe I need to add an (int) cast or (bool) cast as well.

I already change several function locally, (but not is_dir for instance) so more tests pass.

Example of error:
```plaintext
is_dir(): open_basedir restriction in effect. File(/tmp) is not within the allowed path(s): (D:\mdeweerd\workspace\dolibarr\)
```

Here is a list of my current setup:

Tests: 295, Assertions: 29494, Errors: 32, Failures: 14, Warnings: 1, Skipped: 104.

```plaintext
There were 32 errors:

1) NumberingModulesTest::testFactureMercure
Undefined property 'ventil' From D:\mdeweerd\workspace\dolibarr\htdocs\compta\facture\class\facture.class.php:3978.

2) BuildDocTest::testFactureBuild
Error: Using $this when not in object context in D:\mdeweerd\workspace\dolibarr\test\phpunit\BuildDocTest.php:128
Stack trace:
#0 phar://D:/mdeweerd/workspace/dolibarr/phpunit-9.5.phar/phpunit/Framework/TestSuite.php(468): BuildDocTest::setUpBeforeClass()
#1 phar://D:/mdeweerd/workspace/dolibarr/phpunit-9.5.phar/phpunit/Framework/TestSuite.php(508): PHPUnit\Framework\TestSuite->run(Object(PHPUnit\Framework\TestResult))
#2 phar://D:/mdeweerd/workspace/dolibarr/phpunit-9.5.phar/phpunit/TextUI/TestRunner.php(487): PHPUnit\Framework\TestSuite->run(Object(PHPUnit\Framework\TestResult))
#3 phar://D:/mdeweerd/workspace/dolibarr/phpunit-9.5.phar/phpunit/TextUI/Command.php(123): PHPUnit\TextUI\TestRunner->run(Object(PHPUnit\Framework\TestSuite), Array, Array, true)
#4 phar://D:/mdeweerd/workspace/dolibarr/phpunit-9.5.phar/phpunit/TextUI/Command.php(92): PHPUnit\TextUI\Command->run(Array, true)
#5 D:\mdeweerd\workspace\dolibarr\phpunit-9.5.phar(1117): PHPUnit\TextUI\Command::main()
#6 {main}
3) CMailFileTest::testCMailFileHTMLWithImage
is_dir(): open_basedir restriction in effect. File(/tmp) is not within the allowed path(s): (D:\mdeweerd\workspace\dolibarr\)

4) ActionCommTest::testActionCommCreate
Undefined property 'contactid' From D:\mdeweerd\workspace\dolibarr\test\phpunit\ActionCommTest.php:162.  

5) SocieteTest::testSocieteCreate
Undefined property 'actiontypecode' From D:\mdeweerd\workspace\dolibarr\htdocs\core\triggers\interface_50_modAgenda_ActionsAuto.class.php:103.

6) SocieteTest::testSocieteGetFullAddress
Undefined property 'country_code' From D:\mdeweerd\workspace\dolibarr\htdocs\core\class\commonobject.class.php:1041.

7) ReceptionTest::testReceptionCreate
Undefined property 'actiontypecode' From D:\mdeweerd\workspace\dolibarr\htdocs\core\triggers\interface_50_modAgenda_ActionsAuto.class.php:103.

8) ContactTest::testContactOther
Undefined property 'country_code' From D:\mdeweerd\workspace\dolibarr\htdocs\core\class\commonobject.class.php:1041.

9) ContactTest::testContactGetFullAddress
Undefined property 'country_code' From D:\mdeweerd\workspace\dolibarr\htdocs\core\class\commonobject.class.php:1041.

10) AdherentTest::testAdherentTypeCreate
Undefined property 'company' From D:\mdeweerd\workspace\dolibarr\test\phpunit\AdherentTest.php:163.      

11) ProductTest::testProductUpdate
Undefined property 'oldcopy' From D:\mdeweerd\workspace\dolibarr\htdocs\product\class\product.class.php:1169.

12) PricesTest::testUpdatePrice
Undefined property 'ventil' From D:\mdeweerd\workspace\dolibarr\htdocs\compta\facture\class\facture.class.php:3978.

13) DiscountTest::testDiscountCreate
Undefined property 'actiontypecode' From D:\mdeweerd\workspace\dolibarr\htdocs\core\triggers\interface_50_modAgenda_ActionsAuto.class.php:103.

14) ContratTest::testContratCreate
Undefined property 'actiontypecode' From D:\mdeweerd\workspace\dolibarr\htdocs\core\triggers\interface_50_modAgenda_ActionsAuto.class.php:103.

15) FichinterTest::testFichinterCreate
Undefined property 'actiontypecode' From D:\mdeweerd\workspace\dolibarr\htdocs\core\triggers\interface_50_modAgenda_ActionsAuto.class.php:103.

16) CommandeTest::testCommandeCreate
Undefined property 'actiontypecode' From D:\mdeweerd\workspace\dolibarr\htdocs\core\triggers\interface_50_modAgenda_ActionsAuto.class.php:103.

17) CommandeFournisseurTest::testCommandeFournisseurValid
Undefined property 'actiontypecode' From D:\mdeweerd\workspace\dolibarr\htdocs\core\triggers\interface_50_modAgenda_ActionsAuto.class.php:103.

18) FactureTest::testFactureCreate
Undefined property 'ventil' From D:\mdeweerd\workspace\dolibarr\htdocs\compta\facture\class\facture.class.php:3978.

19) FactureRecTest::testFactureRecCreate
Undefined property 'ventil' From D:\mdeweerd\workspace\dolibarr\htdocs\compta\facture\class\facture.class.php:3978.

20) FactureTestRounding::testFactureRoundingCreate1
Undefined property 'total_ht' From D:\mdeweerd\workspace\dolibarr\htdocs\core\class\commonobject.class.php:3779.

21) FactureTestRounding::testFactureAddLine1
Undefined property 'ventil' From D:\mdeweerd\workspace\dolibarr\htdocs\compta\facture\class\facture.class.php:3978.

22) FactureFournisseurTest::testFactureFournisseurCreate
Undefined property 'price' From D:\mdeweerd\workspace\dolibarr\htdocs\fourn\class\fournisseur.facture.class.php:3053.

23) BankAccountTest::testBankAccountFetch
Undefined property 'date_update' From D:\mdeweerd\workspace\dolibarr\htdocs\compta\bank\class\account.class.php:1095.

24) CompanyBankAccountTest::testCompanyBankAccountCreate
Undefined property 'actiontypecode' From D:\mdeweerd\workspace\dolibarr\htdocs\core\triggers\interface_50_modAgenda_ActionsAuto.class.php:103.

25) HolidayTest::testHolidayUpdate
Undefined property 'address' From D:\mdeweerd\workspace\dolibarr\test\phpunit\HolidayTest.php:203.       

26) CategorieTest::testCategorieOther
file_exists(): open_basedir restriction in effect. File(/) is not within the allowed path(s): (D:\mdeweerd\workspace\dolibarr\)

27) KnowledgeRecordTest::testKnowledgeRecordFetch
Undefined property 'fk_c_ticket_category' From D:\mdeweerd\workspace\dolibarr\htdocs\core\class\commonobject.class.php:9621.

28) AccountingAccountTest::testAccountingAccountCreate
Undefined property 'pcg_subtype' From D:\mdeweerd\workspace\dolibarr\test\phpunit\AccountingAccountTest.php:153.

29) WebservicesOtherTest::testWSOtherGetVersions
count(): Parameter must be an array or an object that implements Countable

30) WebservicesUserTest::testWSUserGetUser
count(): Parameter must be an array or an object that implements Countable

31) ModulesTest::testModulesInit
Undefined property 'actiontypecode' From D:\mdeweerd\workspace\dolibarr\htdocs\core\triggers\interface_50_modAgenda_ActionsAuto.class.php:103.

32) EmailCollectorTest::testEmailCollectorCreate
Undefined property 'description' From D:\mdeweerd\workspace\dolibarr\htdocs\core\class\commonobject.class.php:10460.

--

There was 1 warning:

1) LangTest::testLang
assertRegExp() is deprecated and will be removed in PHPUnit 10. Refactor your code to use assertMatchesRegularExpression() instead.

--

There were 32 errors:

1) NumberingModulesTest::testFactureMercure
Undefined property 'ventil' From D:\mdeweerd\workspace\dolibarr\htdocs\compta\facture\class\facture.class.php:3978.

2) BuildDocTest::testFactureBuild
Error: Using $this when not in object context in D:\mdeweerd\workspace\dolibarr\test\phpunit\BuildDocTest.php:128
Stack trace:
#0 phar://D:/mdeweerd/workspace/dolibarr/phpunit-9.5.phar/phpunit/Framework/TestSuite.php(468): BuildDocTest::setUpBeforeClass()
#1 phar://D:/mdeweerd/workspace/dolibarr/phpunit-9.5.phar/phpunit/Framework/TestSuite.php(508): PHPUnit\Framework\TestSuite->run(Object(PHPUnit\Framework\TestResult))
#2 phar://D:/mdeweerd/workspace/dolibarr/phpunit-9.5.phar/phpunit/TextUI/TestRunner.php(487): PHPUnit\Framework\TestSuite->run(Object(PHPUnit\Framework\TestResult))
#3 phar://D:/mdeweerd/workspace/dolibarr/phpunit-9.5.phar/phpunit/TextUI/Command.php(123): PHPUnit\TextUI\TestRunner->run(Object(PHPUnit\Framework\TestSuite), Array, Array, true)
#4 phar://D:/mdeweerd/workspace/dolibarr/phpunit-9.5.phar/phpunit/TextUI/Command.php(92): PHPUnit\TextUI\Command->run(Array, true)
#5 D:\mdeweerd\workspace\dolibarr\phpunit-9.5.phar(1117): PHPUnit\TextUI\Command::main()
#6 {main}
3) CMailFileTest::testCMailFileHTMLWithImage
is_dir(): open_basedir restriction in effect. File(/tmp) is not within the allowed path(s): (D:\mdeweerd\workspace\dolibarr\)

4) ActionCommTest::testActionCommCreate
Undefined property 'contactid' From D:\mdeweerd\workspace\dolibarr\test\phpunit\ActionCommTest.php:162.  

5) SocieteTest::testSocieteCreate
Undefined property 'actiontypecode' From D:\mdeweerd\workspace\dolibarr\htdocs\core\triggers\interface_50_modAgenda_ActionsAuto.class.php:103.

6) SocieteTest::testSocieteGetFullAddress
Undefined property 'country_code' From D:\mdeweerd\workspace\dolibarr\htdocs\core\class\commonobject.class.php:1041.

7) ReceptionTest::testReceptionCreate
Undefined property 'actiontypecode' From D:\mdeweerd\workspace\dolibarr\htdocs\core\triggers\interface_50_modAgenda_ActionsAuto.class.php:103.

8) ContactTest::testContactOther
Undefined property 'country_code' From D:\mdeweerd\workspace\dolibarr\htdocs\core\class\commonobject.class.php:1041.

9) ContactTest::testContactGetFullAddress
Undefined property 'country_code' From D:\mdeweerd\workspace\dolibarr\htdocs\core\class\commonobject.class.php:1041.

10) AdherentTest::testAdherentTypeCreate
Undefined property 'company' From D:\mdeweerd\workspace\dolibarr\test\phpunit\AdherentTest.php:163.      

11) ProductTest::testProductUpdate
Undefined property 'oldcopy' From D:\mdeweerd\workspace\dolibarr\htdocs\product\class\product.class.php:1169.

12) PricesTest::testUpdatePrice
Undefined property 'ventil' From D:\mdeweerd\workspace\dolibarr\htdocs\compta\facture\class\facture.class.php:3978.

13) DiscountTest::testDiscountCreate
Undefined property 'actiontypecode' From D:\mdeweerd\workspace\dolibarr\htdocs\core\triggers\interface_50_modAgenda_ActionsAuto.class.php:103.

14) ContratTest::testContratCreate
Undefined property 'actiontypecode' From D:\mdeweerd\workspace\dolibarr\htdocs\core\triggers\interface_50_modAgenda_ActionsAuto.class.php:103.

15) FichinterTest::testFichinterCreate
Undefined property 'actiontypecode' From D:\mdeweerd\workspace\dolibarr\htdocs\core\triggers\interface_50_modAgenda_ActionsAuto.class.php:103.

16) CommandeTest::testCommandeCreate
Undefined property 'actiontypecode' From D:\mdeweerd\workspace\dolibarr\htdocs\core\triggers\interface_50_modAgenda_ActionsAuto.class.php:103.

17) CommandeFournisseurTest::testCommandeFournisseurValid
Undefined property 'actiontypecode' From D:\mdeweerd\workspace\dolibarr\htdocs\core\triggers\interface_50_modAgenda_ActionsAuto.class.php:103.

18) FactureTest::testFactureCreate
Undefined property 'ventil' From D:\mdeweerd\workspace\dolibarr\htdocs\compta\facture\class\facture.class.php:3978.

19) FactureRecTest::testFactureRecCreate
Undefined property 'ventil' From D:\mdeweerd\workspace\dolibarr\htdocs\compta\facture\class\facture.class.php:3978.

20) FactureTestRounding::testFactureRoundingCreate1
Undefined property 'total_ht' From D:\mdeweerd\workspace\dolibarr\htdocs\core\class\commonobject.class.php:3779.

21) FactureTestRounding::testFactureAddLine1
Undefined property 'ventil' From D:\mdeweerd\workspace\dolibarr\htdocs\compta\facture\class\facture.class.php:3978.

22) FactureFournisseurTest::testFactureFournisseurCreate
Undefined property 'price' From D:\mdeweerd\workspace\dolibarr\htdocs\fourn\class\fournisseur.facture.class.php:3053.

23) BankAccountTest::testBankAccountFetch
Undefined property 'date_update' From D:\mdeweerd\workspace\dolibarr\htdocs\compta\bank\class\account.class.php:1095.

24) CompanyBankAccountTest::testCompanyBankAccountCreate
Undefined property 'actiontypecode' From D:\mdeweerd\workspace\dolibarr\htdocs\core\triggers\interface_50_modAgenda_ActionsAuto.class.php:103.

25) HolidayTest::testHolidayUpdate
Undefined property 'address' From D:\mdeweerd\workspace\dolibarr\test\phpunit\HolidayTest.php:203.       

26) CategorieTest::testCategorieOther
file_exists(): open_basedir restriction in effect. File(/) is not within the allowed path(s): (D:\mdeweerd\workspace\dolibarr\)

27) KnowledgeRecordTest::testKnowledgeRecordFetch
Undefined property 'fk_c_ticket_category' From D:\mdeweerd\workspace\dolibarr\htdocs\core\class\commonobject.class.php:9621.

28) AccountingAccountTest::testAccountingAccountCreate
Undefined property 'pcg_subtype' From D:\mdeweerd\workspace\dolibarr\test\phpunit\AccountingAccountTest.php:153.

29) WebservicesOtherTest::testWSOtherGetVersions
count(): Parameter must be an array or an object that implements Countable

30) WebservicesUserTest::testWSUserGetUser
count(): Parameter must be an array or an object that implements Countable

31) ModulesTest::testModulesInit
Undefined property 'actiontypecode' From D:\mdeweerd\workspace\dolibarr\htdocs\core\triggers\interface_50_modAgenda_ActionsAuto.class.php:103.

32) EmailCollectorTest::testEmailCollectorCreate
Undefined property 'description' From D:\mdeweerd\workspace\dolibarr\htdocs\core\class\commonobject.class.php:10460.

--

There was 1 warning:

1) LangTest::testLang
assertRegExp() is deprecated and will be removed in PHPUnit 10. Refactor your code to use assertMatchesRegularExpression() instead.

There were 14 failures:

1) FunctionsLibTest::testIsValidMXRecord
Failed asserting that 1 matches expected 0.

2) CodingPhpTest::testPHP  - IGNORE this, this was a local leftover (maybe the test should limit itself to tracked git files).
Found a tag <br /> that is for xml in file admin/dolistore/class/PSWebServiceLibrary.php.orig. You must use html syntax <br> instead.
Failed asserting that false is true.

3) SocieteTest::testSocieteMerge
Failed asserting that 0 is equal to -3 or is less than -3.

4) PropalTest::testPropalValid
Failed asserting that 0 is less than -1.

5) RestAPIUserTest::testRestGetUser
Parsing of json result must not be null
Failed asserting that null is not null.

6) RestAPIContactTest::testRestGetContact
Parsing of json result must not be null
Failed asserting that null is not null.

7) RestAPIDocumentTest::testPushDocument
Parsing of json result must not be null
Failed asserting that null is not null.

8) WebservicesProductsTest::testWSProductsCreateProductOrService
Failed asserting that null matches expected 'OK'.

9) WebservicesInvoicesTest::testWSInvoicesCreateInvoice
Failed asserting that null matches expected 'OK'.

10) WebservicesOrdersTest::testWSOrderGetOrder
Failed asserting that null matches expected 'OK'.

11) WebservicesThirdpartyTest::testWSThirdpartycreateThirdParty
Failed asserting that null matches expected 'OK'.

12) ScriptsTest::testBank
Failed asserting that two strings are equal.
--- Expected
+++ Actual
@@ @@
-''
+'Failed to find bank account with ref BANKDUMMY.'

13) PaypalTest::testPaypalOk
Failed asserting that 404 matches expected 200.

14) StripeTest::testStripeOk


ERRORS!
Tests: 295, Assertions: 29494, Errors: 32, Failures: 14, Warnings: 1, Skipped: 104.
```

mdeweerd commented:
Most of the phpunit test results are now also available Windows CI by catching the Security Exceptions that are generated when simply including the `main.inc.php` when the testcase files are included (before the actual tests).

mdeweerd commented:
The win-ci workflow now attaches an artefact containing the files `documents/dolibarr.log` and the output from php_unit as `phpunit_tests.log`.

We now have the following results with the base code (excluding tests that are skipped because of main.inc.php):
Without DolDeprecatedHandler (see #28187)
```plaintext
Tests: 366, Assertions: 77417, Errors: 7, Failures: 12, Skipped: 15.

With DolDeprecatedHandler (increase related to the use of deprecated and undefined properties):
```plaintext
Tests: 366, Assertions: 77105, Errors: 37, Failures: 13, Skipped: 95.
```

eldy commented on htdocs/core/class/commonobject.class.php:
```diff
@@ -783,6 +811,27 @@ abstract class CommonObject
+		return array(
+			'alreadypaid' => 'totalpaid',
+			'cond_reglement' => 'cond_reglement_id',
+			'modelpdf' => 'model_pdf',
```
This field has been definitely removed from develop branch so we can remove it here too.

mdeweerd replied:
@eldy The question is : do you want to provide some backwards compatibility or not.
This table is the mapping in case an access to the old value is made.

If you want me to make deprecated fields unsupported, let me know which ones to remove - I am pretty sure that some need to be kept.

mdeweerd commented:
Currently:
```
Tests: 2681, Assertions: 106045, Errors: 28, Failures: 5, Skipped: 82.
```
Before:
```
Tests: 380, Assertions: 89519, Errors: 34, Failures: 5, Skipped: 87.
```

eldy commented on htdocs/core/class/commonobject.class.php:
```diff
@@ -320,11 +341,11 @@ abstract class CommonObject
+	 * @see $status
 	 */
-	public $statut;
+	private $statut;
```
A lot of field are deprecated since a short time, so we can't remove their usability too quickly. If property is set to private, all external modules and custom dev will be broken without having enough time for a compatibility period.

mdeweerd replied:
> A lot of field are deprecated since a short time, so we can't remove their usability too quickly. If property is set to private, all external modules and custom dev will be broken without having enough time for a compatibility period.

The fields still exist as virtual properties - the private fields are only there to help the IDE - they were added for that purpose.

The DolDeprecationHandler will serve the '$status' property when '$statut' is accessed and update the '$status' property when '$statut' is updated.  It avoid any need to have double assignments in the code where we see "$this->status = $something; $this->statut = $something;" .
IMHO these double assignments are easily forgotten but with this Trait that would no longer be an issue.

At the same time as the deprecated properties are virtual any access is captured and reported as a deprecated access.

The PHPUNIT test demonstrates this: - one example where the oldPropertyName is assigned and the value is read back through the new property name:

https://github.com/Dolibarr/dolibarr/blob/f2001cb05449161d3b7cf61f0004ef6d1a602e06/test/phpunit/DolDeprecationHandlerTest.php#L241-L248

eldy replied:
I don't understand. 
If the code of an external module do:
$theobject->statut = 1;
On any object that etends commobject.class.php (so any business object), if property statut is set as private, this will generate a php warning/error. Did i miss something.

mdeweerd replied:
The deprecation warnings are only generated when the deprecation detection is on.
So that is not on in a production environment and should not make things fail.

The "trigger_error" calls are gated with 
https://github.com/Dolibarr/dolibarr/blob/492efeeaacdd3de71c21cd0632dd52a4c97d0d2c/htdocs/core/class/doldeprecationhandler.class.php#L158-L161

and the value return by an optional function in the class the Trait is attached to:
`isDynamicPropertiesEnabled()`.

When the latter is set, access to dynamic properties trigger a notice.

In a test setup the deprecation warnings are active and certain accesses will result in a failing test.

This can be observed in the current test results.

There can be several steps: 
1. Integrate DolDeprecatedHandler with "isDynamicPropertiesEnabled()" returning "true" by default (the tests should pass);
2. Create a new PR with isDynamicPropertiesEnabled() returning false by default (the same tests will fail);
3. In the futur attach DolDeprecatedHandler progressively to other classes/extend the deprecation mapping in subclasses and remove the deprecated definitions and explicit assignments to thos definitions.

mdeweerd commented:
I modified the default behavior with regards to accessing dynamic properties that do not exist, fixed some changes to conf.class.php, and added logging using dol_syslog.

Here is where the default behavior is set for accessing dynamimc properties, I also explained the options.

https://github.com/Dolibarr/dolibarr/blob/6e5eee82e93fc3b2a6558204a85e711fd500259a/htdocs/core/class/doldeprecationhandler.class.php#L184-L203

mdeweerd commented:
@eldy: The test case FactureTest::testFactureUpdate fails.

It seems that the code "supposes" that $this->cond_reglement_id is not a string, so the contents is not quoted, resulting in inserting the following in the query:
`, fk_cond_reglement=A réception, fk_mode_reglement=0,`

from here

https://github.com/Dolibarr/dolibarr/blob/ff0eea8ce6bbc3426fbf23012a9cdd988a806f79/htdocs/compta/facture/class/facture.class.php#L2521

I am checking how cond_reglement_id got the wrong value but the above may need consideration as far as the approach to protecting from sql insertion is concerned.
